### PR TITLE
Add compatibility tests for v5.0.x #2

### PR DESCRIPTION
**Pull request by chatton**
## Description

This PR replaces #2370

This is a re-work of the referenced PR. Instead of using a large json file per release branch, this method allows us to define smaller, more maintainable files per release branch / test suite pair.

Sample workflow: [here](https://github.com/cosmos/ibc-go/actions/runs/3129300394)

Update: One additional benefit, each job is limited to creating 256 tasks. A github workflow can have multiple jobs, i.e. transfer, client, incentivized etc. While we are not at that level yet (transfer test is creating 80), splitting up  like this leaves more room for growth without refactoring how the tests run in the future.

closes #2370

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/master/CONTRIBUTING.md#testing)
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [x] Re-reviewed `Files changed` in the Github PR explorer
- [ ] Review `Codecov Report` in the comment section below once CI passes

### CODE REVIEW

File: .github/workflows/e2e-compatibility-workflow-call.yaml [hunk 1]
```diff
@@ -0,0 +1,50 @@
+on:
+  workflow_call:
```
Comment (chatton) [inline .github/workflows/e2e-compatibility-workflow-call.yaml:2]: this is a re-usable workflow that we can call for each test suite.

**Comment by chatton** [comment]
@colin-axner & @charleenfei thank you for reviewing #2370 after some discussion with @crodriguezvega and looking some more into Colin's suggestion, I refactored the workflow to be (much) more maintainable. The workflows are largely the same, it's mostly just the smaller config file we need to manage.

File: .github/workflows/e2e-compatibility.yaml [hunk 2]
```diff
@@ -47,42 +48,29 @@ jobs:
...
-          # to create the matrix in the following job.
-          test_matrix="$(cat .github/compatibility-test-matrices/${{ needs.determine-docker-tag.outputs.docker-tag }}.json | jq -c)"
-          echo $test_matrix
-          echo "::set-output name=test-matrix::$test_matrix"
-        id: set-test-matrix
+  transfer:
+    needs:
+      - build-release-image
+      - determine-docker-tag
+    uses: ./.github/workflows/e2e-compatibility-workflow-call.yaml
```
Comment (chatton) [inline .github/workflows/e2e-compatibility.yaml:55]: this is how workflows can be re-used in github actions.

File: .github/workflows/e2e-compatibility.yaml [hunk 3]
```diff
@@ -47,42 +48,29 @@ jobs:
...
+    with:
+      docker-tag: "${{ needs.determine-docker-tag.outputs.docker-tag }}"
+      test-suite: "transfer"
```
Comment (chatton) [inline .github/workflows/e2e-compatibility.yaml:58]: these strings map to sub directories, I'm open to alternatives if there are any suggestions!

Comment (damiannolan) [reply .github/workflows/e2e-compatibility.yaml:58]: Will we also need to add blocks for `interchain-accounts` here?

File: .github/workflows/e2e-compatibility-workflow-call.yaml [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+          echo $test_matrix
+          echo "::set-output name=test-matrix::$test_matrix"
+        id: set-test-matrix
+
+  e2e:
+    runs-on: ubuntu-latest
+    needs: load-test-matrix
+    strategy:
+      fail-fast: false
+      matrix: ${{ fromJSON(needs.load-test-matrix.outputs.test-matrix) }}
```
Comment (damiannolan) [inline .github/workflows/e2e-compatibility-workflow-call.yaml:34]: Aha, so this magic uses `load-test-matrix` to parse the JSON files and set the values in `matrix` which is used below to run the e2e test and populate the env?

**Review by damiannolan — APPROVED** [review]
Just left a few questions, I'm getting a better grasp on the workflows, slowly but surely its becoming easier to parse.

The yaml gods have spoken! ⚒️⚡ 

L: 
  G:
    T:
      M: 

❤️

File: .github/workflows/e2e-compatibility.yaml [hunk 5]
```diff
@@ -1,12 +1,13 @@
...
-      type: choice
-      options:
-        - release/v5.0.x
+    inputs:
+      release-branch:
+        description: 'Release branch to test'
+        required: true
+        type: choice
+        options:
+          - release/v5.0.x
```
[the remark this answers is not in the document]
Comment (chatton) [reply .github/workflows/e2e-compatibility.yaml:10]: yes - this is just for a dropdown, if we prefer, we can have a text box and enter anything we like. I just think the dropdown is better UX

File: .github/workflows/e2e-compatibility-workflow-call.yaml (hunk 4 above)
Comment (chatton) [reply .github/workflows/e2e-compatibility-workflow-call.yaml:34]: yes the previous job `load-test-matrix` stores a json string as its output (the contents of a json file depending on args) and the `matrix` field accepts dynamic input (or statically defined in the workflow)

**Comment by chatton** [comment]
@damiannolan 

> Will we also need to add blocks for interchain-accounts here?

Currently all the interchain accounts tests are running using the icad, so the state of the release branch in `ibc-go` should have no impact here right?

> I'm not sure I understand fully here. Could you provide me with the verbose output

Right maybe I can add some clarification, the tests themselves exist under `e2e/...` but the matrix files for the workflow live separately (the new dir added in this PR) `compatibility-test-matrices`. 

This additional field e.g. `test-suite: "transfer"` is literally just a convention I'm using to pass arguments to the reusable workflow so that we can run any test suite the same way by adhering to a directory structure.

`compatibility-test-matrices/<release-<version>/<test-suite>.json`

File: .github/workflows/e2e-compatibility.yaml (hunk 5 above)
[the remark this answers is not in the document]
Comment (damiannolan) [reply .github/workflows/e2e-compatibility.yaml:10]: Agree!

File: .github/workflows/e2e-compatibility.yaml (hunk 3 above)
Comment (chatton) [reply .github/workflows/e2e-compatibility.yaml:58]: maybe we could rename the value to be something explicit like `matrix-filename` WDYT?

Comment (damiannolan) [reply .github/workflows/e2e-compatibility.yaml:58]: I'm easy. I think its fine to leave as is

File: .github/workflows/e2e-compatibility-workflow-call.yaml (hunk 4 above)
Comment (colin-axner) [reply .github/workflows/e2e-compatibility-workflow-call.yaml:34]: This is indeed magic :sweat_smile:

File: .github/workflows/e2e-compatibility.yaml (hunk 3 above)
Comment (colin-axner) [reply .github/workflows/e2e-compatibility.yaml:58]: Whatever is fine by me!

File: .github/compatibility-test-matrices/release-v5.0.x/client.json [hunk 6]
```diff
@@ -0,0 +1,10 @@
+{
+  "chain-a": ["release-v5.0.x", "v4.1.0", "v3.3.0", "v2.4.0"],
+  "chain-b": ["release-v5.0.x", "v4.1.0", "v3.3.0", "v2.4.0"],
```
Comment (colin-axner) [inline .github/compatibility-test-matrices/release-v5.0.x/client.json:3]: so clean! :heart:

**Review by colin-axner — APPROVED** [review]
absolutely fabulous! :crown: 

These e2e test additions are like a dream come true!

**Comment by colin-axner** [comment]
> Currently all the interchain accounts tests are running using the icad, so the state of the release branch in ibc-go should have no impact here right?

I guess the question is if we want to be able to trigger icad compatibility tests from ibc-go as well (long term no, but maybe in the short term?)

**Comment by chatton** [comment]
@colin-axner 

> I guess the question is if we want to be able to trigger icad compatibility tests from ibc-go as well (long term no, but maybe in the short term?)

Right, we can already do this with the `Manual E2E (Icad)` test. But I think long term will be able to use simd and it can be added as a new file here.

File: .github/workflows/e2e-compatibility-workflow-call.yaml (hunk 4 above)
Comment (chatton) [reply .github/workflows/e2e-compatibility-workflow-call.yaml:34]: the process of passing variables between steps is quite cumbersome unfortunately 😓

**Pull request merged**